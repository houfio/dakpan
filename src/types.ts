import { ComponentType, FunctionComponent } from 'react';

export type Dakpan<S, A extends Actions<S>> = [FunctionComponent, DakpanHook<S, A>, DakpanHoc<S, A>];

export type DakpanHook<S, A extends Actions<S>> = () => [S, MappedActions<S, A>];

export type DakpanHoc<S, A extends Actions<S>> = <H>(
  map: (state: S, actions: MappedActions<S, A>) => H
) => <P>(
  component: ComponentType<P & H>
) => ComponentType<P>;

export type DakpanContext<S, A extends Actions<S>> = undefined | {
  state: S,
  actions: MappedActions<S, A>
};

export type Actions<S> = {
  [action: string]: (...args: any[]) => (state: S) => S | undefined | Promise<S | undefined>
};

export type MappedActions<S, A extends Actions<S>> = {
  [K in keyof A]: ((...args: Parameters<A[K]>) => Promise<void>) & {
    c: (...args: Parameters<A[K]>) => () => Promise<void>
  }
};

export type InitialState<S> = S | (() => S);
