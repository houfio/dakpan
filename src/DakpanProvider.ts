import { createElement, ReactElement } from 'react';

import { DakpanMultiProviderProps } from './types';

export const DakpanProvider = ({ provide, children }: DakpanMultiProviderProps) => (
  provide.reduce((previous, current) => createElement(current, { children: previous }), children as ReactElement<any>)
);
