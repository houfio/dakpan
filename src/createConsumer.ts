import { ComponentClass, Consumer, ConsumerProps, createElement, ReactElement, StatelessComponent } from 'react';

import { Actions, DakpanConsumerProps, MappedActions } from './types';

export const createConsumer = <S, A extends Actions<S>>(
  Consumer: Consumer<S>,
  actions: MappedActions<S, A>
) => ({ children }: DakpanConsumerProps<S, A>) => createElement(Consumer, {
  children: (value: S) => children(value, actions)
});
