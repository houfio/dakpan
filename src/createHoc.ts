import { createElement, FunctionComponent } from 'react';

import { Actions, DakpanHoc, DakpanHook } from './types';

export const createHoc = <S, A extends Actions<S>>(
  hook: DakpanHook<S, A>
): DakpanHoc<S, A> => (map) => (component) => {
  const hoc: FunctionComponent<any> = (props) => {
    const [state, actions] = hook();

    return createElement(component, {
      ...props,
      ...map(state, actions)
    });
  };

  hoc.displayName = `WithDakpan(${component.displayName || component.name || 'Unknown'})`;

  return hoc;
};
