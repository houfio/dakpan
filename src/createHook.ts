import { Context, useContext } from 'react';

import { Actions, DakpanContext, DakpanHook } from './types';

export function createHook<S, A extends Actions<S>>(context: Context<DakpanContext<S, A>>): DakpanHook<S, A> {
  return () => {
    const state = useContext(context);

    if (!state) {
      throw Error('Provider not mounted');
    }

    return [
      state.state,
      state.actions
    ];
  };
}
