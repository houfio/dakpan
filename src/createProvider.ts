import { Context, createElement, FunctionComponent, useEffect, useMemo, useRef, useState } from 'react';

import { Actions, DakpanContext, MappedActions } from './types';

export function createProvider<S, A extends Actions<S>>(
  context: Context<DakpanContext<S, A>>,
  actions: A,
  state?: S
): FunctionComponent<any> {
  return ({ children, value, reinitialize = false }) => {
    const [, update] = useState({});
    const initialRef = useRef(value);
    const stateRef = useRef(state || value);
    const mapped = useMemo(() => Object.keys(actions).reduce(
      (previous, current) => {
        const execute = (...args: unknown[]) => () => Promise.resolve(actions[current](...args)(stateRef.current))
          .then((next) => {
            if (next) {
              stateRef.current = next;
              update({});
            }
          });

        const action = (...args: unknown[]) => execute(...args)();
        action.c = execute;

        return {
          ...previous,
          [current]: action
        };
      },
      {}
    ) as MappedActions<S, A>, [actions, stateRef]);

    useEffect(() => {
      if (reinitialize && value !== initialRef.current) {
        stateRef.current = initialRef.current = value;
        update({});
      }
    }, [value, reinitialize, initialRef]);

    return createElement(context.Provider, {
      value: {
        state: stateRef.current,
        actions: mapped
      }
    }, children);
  };
}
