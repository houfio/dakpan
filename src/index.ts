export { createDakpan } from './createDakpan';
export { DakpanProvider } from './DakpanProvider';
