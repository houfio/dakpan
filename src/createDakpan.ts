import {
  Component,
  ComponentClass,
  ConsumerProps,
  createContext,
  ProviderProps,
  ReactElement,
  ReactNodeArray,
  ReactPortal,
  StatelessComponent
} from 'react';

import { createConsumer } from './createConsumer';
import { createProvider } from './createProvider';
import { Actions, DakpanProviderProps, GetState, MappedActions, SetState } from './types';
import { withDakpan } from './withDakpan';

export const createDakpan = <S>(initialState: S) => <A extends Actions<S>>(actions: A) => {
  let getState: GetState<S> | undefined;
  let setState: SetState<S, DakpanProviderProps> | undefined;

  const mergeAction = (result: Partial<S>) => {
    if (!getState || !setState) {
      throw new Error('Provider unmounted before action could complete');
    } else if (typeof result !== 'object') {
      throw new Error('Actions may only return objects');
    }

    const newState = {
      ...getState() as any,
      ...result as any
    };

    setState(newState);

    return newState;
  };

  const { Provider, Consumer } = createContext<S>(initialState);
  const mappedActions = Object.keys(actions).reduce(
    (result, key) => {
      const e = (...args: any[]) => () => {
        const action = actions[key];

        if (!getState || !setState) {
          throw new Error('You may not dispatch an action before its provider is mounted');
        }

        const state = getState();
        const result = action(...args)(state);

        if (result instanceof Promise) {
          return result.then(mergeAction);
        }

        return mergeAction(result);
      };

      const action: any = (...args: any[]) => e(...args)();
      action.e = e;

      return {
        ...result,
        [key]: action
      };
    },
    {}
  ) as MappedActions<S, A>;

  const provider = createProvider(Provider, initialState, (get, set) => {
    getState = get;
    setState = set;
  });
  const consumer = createConsumer(Consumer, mappedActions);

  return {
    Provider: provider,
    Consumer: consumer,
    actions: mappedActions,
    withDakpan: withDakpan(consumer)
  };
};
