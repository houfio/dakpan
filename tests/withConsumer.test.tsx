import * as React from 'react';
import { render } from 'react-testing-library';

import { createDakpan, DakpanProvider, withDakpan } from '../src';

it('should wrap the component with specified hocs', () => {
  const { Provider: FirstProvider, withConsumer: withFirst } = createDakpan({
    hello: 'world'
  })({});

  const { Provider: SecondProvider, withConsumer: withSecond } = createDakpan({
    test: 234
  })({});

  const Component = withDakpan(withFirst((state) => state), withSecond((state) => state))(({ hello, test }) => (
    <>
      {hello}
      {test}
    </>
  ));

  const { asFragment } = render(
    <DakpanProvider provide={[FirstProvider, SecondProvider]}>
      <Component/>
    </DakpanProvider>
  );

  expect(asFragment()).toMatchSnapshot();
});

it('should return the component when no hocs are specified', () => {
  const Component = withDakpan()(() => (
    <div/>
  ));

  const { asFragment } = render(
    <Component/>
  );

  expect(asFragment()).toMatchSnapshot();
});
