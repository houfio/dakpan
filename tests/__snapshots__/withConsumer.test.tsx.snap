// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return the component when no hocs are specified 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`should wrap the component with specified hocs 1`] = `
<DocumentFragment>
  world234
</DocumentFragment>
`;
